docs/rd-gen.md

```diff
@@ -0,0 +1,249 @@
+```
+
+This root object is a top-level model, and thus can contain members (see below), but it's also possible to add extensions on top of it:
+
+```kotlin
+object ExampleExt : Ext(ExampleRootNova)
+```
+
+This extension is a separate model (you may imagine them like namespaces with an additional ability to hold members with static lifetime).
+
+You may put everything from your protocol into a root model, or define separate extensions for product subsystems and such. 
+
+### Types
+Before we discuss the model members, let's discuss the member types. RdGen allows the user to define their own types that then may be used in models (and may be nested into each other). You define a type via several ways:
+
+1. By calling a corresponding method on a top-level object, for example:
+   ```kotlin
+   object Example : Root() {
+      val nestedType = structdef {
+          // Nested members of the type go here
+      }
+   }
+   ```
+   
+   This defines a struct named `nestedType` inside of an `Example` model.
+2. By calling a method in a place where a corresponding type is expected, in an ad-hoc manner, for example:
+   ```kotlin
+   object Example : Root() {
+      init {
+          field("foobar", structdef("OptionalNameForThisType") {
+              // Nested members of the type go here
+          })
+      }
+   }
+   ```
+   
+You can reuse a same type declared once in several places (for example, as a type for several fields or a return type of a call).
+
+#### Type Categories
+Most of the types from the coming section are split into two big categories: _bindable types_ and _scalars_.
+
+A bindable type is a type that can contain reactive members, while a scalar type is immutable and cannot contain any reactive or bindable members in it.
+
+#### Entity Types
+The protocol allows to declare the following type of entities:
+1. `structdef`: a simple _non-bindable type_ that cannot contain any reactive properties; only fields are allowed. Use structs for simple data, normally for value objects only.
+   
+   Structs support inheritance, use `basestruct` and `extends` construct if you need:
+   ```kotlin
+   val BaseStruct = basestruct {
+       field("baseField", PredefinedType.int)
+   }
+   val DerivedStruct = structdef extends BaseStruct {
+       field("derivedField", PredefinedType.bool)
+   }
+   ```
+2. `classdef`: a bindable type that may contain any reactive members inside. Classes also support inheritance:
+   ```kotlin
+   val BaseClass = baseclass {
+       field("baseField", PredefinedType.int)
+   }
+   val DerivedClass = classdef extends BaseClass {
+       field("derivedField", PredefinedType.bool)
+   }
+   ```
+
+   Classes may be open for external inheritance as well. For that, use `openclass` instead of `classdef`.
+3. `interfacedef`: declares an interface that may be implemented by other types. Example:
+   ```kotlin
+   val Interface = interfacedef {
+       call("foo", PredefinedType.void, PredefinedType.void)
```

Comment:
I'm not sure it works. Because we don't support fields for interfaces. How is `call` supposed to work with structures? I think it's a bug that we allow calls here. 
As far as I know, we only support `method` here, but it's dark magic that we don't use it.